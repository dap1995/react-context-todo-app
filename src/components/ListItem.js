// @flow
import React from 'react';
import styled, { css } from 'styled-components';
import AcceptIcon from '../accept-icon.png';

const Text = styled.span`
  display: flex;
  flex: 8;
  flex-wrap: nowrap;
  padding: 10px;
  transition: color 0.3s linear 0.2s;
  font-size: 24px;
  overflow: scroll;

  ${({ checked }) =>
    checked &&
    css`
      text-decoration: line-through;
      color: #ddd;
      transition: all 0.3s linear;
    `};
`;

const Item = styled.li`
  display: flex;
  flex: 1;
  flex-wrap: wrap;
  overflow: hidden;
  flex-direction: row;
  list-style: none;
  border-bottom: 1px solid #ddd;
  align-items: center;

  &:last-child {
    border-bottom: 0;
  }
`;

const CheckBox = styled.button`
  display: flex;
  flex: 1;
  background-color: white;
  margin: 0.5em;
  min-width: 40px;
  min-height: 40px;
  max-width: 40px;
  max-height: 40px;
  border-radius: 100%;
  background-size: 0%;
  background-image: url(${AcceptIcon});
  background-repeat: no-repeat;
  background-position: center;
  transition: background-size 0.4s cubic-bezier(0.06, 1.74, 0.65, -1.14),
    background-color 0.4s ease 0.2s;

  ${({ checked }) =>
    checked &&
    css`
      background-color: rgba(175, 47, 47, 0.5);
      background-size: 50%;
      transition: background-size 0.4s cubic-bezier(0.6, -1, 0.7, 2),
        background-color 0.4s ease;
    `};
`;

const RemoveButton = styled.button`
  display: flex;
  flex: 1;
  justify-content: center;
  min-width: 10px;
  max-width: 10px;
  margin: 0.5em;
  background-color: transparent;
  border: 0;
  font-size: 24px;
  font-weight: 300;
  color: transparent;
  transition: all 0.2s linear;

  ${Item}:hover & {
    color: #cc9a9a;
    transition: all 0.2s linear;
  }

  ${Item}:hover &:hover {
    color: #af5b5e;
  }
`;

type Props = {
  text: string,
  id: string,
  onClose: (index: string) => void,
  onRemove: (index: string) => void,
  checked: boolean
};

/**
 * Represents an Item Component from a List Component.
 */
class ListItem extends React.PureComponent<Props> {
  onClose = () => {
    const { id, onClose } = this.props;
    onClose(id);
  };

  onRemove = () => {
    const { id, onRemove } = this.props;
    onRemove(id);
  };

  render() {
    const { text, checked } = this.props;
    return (
      <Item>
        <CheckBox onClick={this.onClose} checked={checked} />
        <Text checked={checked}>{text}</Text>
        <RemoveButton onClick={this.onRemove}>{'X'}</RemoveButton>
      </Item>
    );
  }
}

export default ListItem;
