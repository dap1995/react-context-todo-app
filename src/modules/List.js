/**
 * This module provides useful pure functions
 * to obtain information from a todo List.
 */
import { VisibilityFilter } from './Visibility';

export function filterListByVisibility(list = [], visibility) {
  switch (visibility) {
    case VisibilityFilter.DONE:
      return list.filter(todo => todo.isDone);

    case VisibilityFilter.TODO:
      return list.filter(todo => !todo.isDone);

    default:
      return list;
  }
}

export function pluralize(count, word) {
  return count !== 1 ? `${word}s` : word;
}
