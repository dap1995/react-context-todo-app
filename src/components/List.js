// @flow
import * as React from 'react';
import styled from 'styled-components';
import memoize from 'memoize-one';
import ListItem from './ListItem';
import Option from './MenuOption';
import type { Todo } from '../flowtypes';
import {
  VisibilityFilter as Visibility,
  filterListByVisibility,
  pluralize
} from '../modules/List';
import Button from './Button';

const Container = styled.div`
  display: flex;
  flex: 1;
  flex-wrap: nowrap;
  flex-direction: column;
`;

const Content = styled.div`
  display: flex;
  flex: 1;
  flex-wrap: nowrap;
  flex-direction: column;
  overflow: hidden;
`;

const FilterMenu = styled.div`
  display: flex;
  flex: 8;
  flex-wrap: wrap;
  justify-content: center;
  align-items: space-around;
`;

const Footer = styled.div`
  display: flex;
  flex: 1;
  flex-direction: row;
  flex-wrap: wrap;
  border-top: 1px solid #e6e6e6;
  box-shadow: 0 1px 1px rgba(0, 0, 0, 0.2), 0 8px 0 -3px #f6f6f6,
    0 9px 1px -3px rgba(0, 0, 0, 0.2), 0 16px 0 -6px #f6f6f6,
    0 17px 2px -6px rgba(0, 0, 0, 0.2);
`;

const Info = styled.span`
  display: flex;
  flex: 1;
  margin: 5px;
  flex-wrap: wrap;
  align-items: center;
  justify-content: flex-start;
  color: #bfbfbf;
  font-size: 12px;
  font-weight: 300;
`;

const ClearButton = Button.extend`
  border: none;
  justify-content: flex-end;

  &:hover {
    text-decoration: underline;
  }
`;

type Props = {
  list: Todo[],
  visibility: string,
  onEditItem: (item: Todo) => void,
  onRemoveItem: (index: string) => void,
  onCloseItem: (index: string) => void,
  onClear: () => void
};

type State = {
  list: Todo[]
};

/**
 * List component to render todo items
 */
export default class List extends React.PureComponent<Props, State> {
  filter = memoize((list, visibility) =>
    filterListByVisibility(list, visibility)
  );

  hasCompleted = () => {
    const { list } = this.props;
    return list.find(item => item.isDone);
  };

  countItemsLeft = () => {
    const { list } = this.props;
    return list.reduce((acc, { isDone }) => (!isDone ? acc + 1 : acc), 0);
  };

  render() {
    const {
      list,
      visibility,
      onEditItem,
      onRemoveItem,
      onCloseItem,
      onClear
    } = this.props;
    const filteredList = this.filter(list, visibility);
    const itemsLeft = this.countItemsLeft();
    return (
      <Container>
        <Content>
          {filteredList.map(item => (
            <ListItem
              id={item.id}
              key={`${item.id}`}
              text={item.text}
              onEdit={onEditItem}
              onClose={onCloseItem}
              onRemove={onRemoveItem}
              checked={item.isDone}
            />
          ))}
        </Content>
        <Footer>
          <Info>{`${itemsLeft} ${pluralize(itemsLeft, 'item')} left`}</Info>
          <FilterMenu>
            <Option href={'#'} enabled={visibility === Visibility.NONE}>
              {'All'}
            </Option>
            <Option
              href={`#/${Visibility.TODO}`}
              enabled={visibility === Visibility.TODO}
            >
              {'Active'}
            </Option>
            <Option
              href={`#/${Visibility.DONE}`}
              enabled={visibility === Visibility.DONE}
            >
              {'Completed'}
            </Option>
          </FilterMenu>
          <ClearButton onClick={onClear} hidden={!this.hasCompleted()}>
            {'Clear completed'}
          </ClearButton>
        </Footer>
      </Container>
    );
  }
}
