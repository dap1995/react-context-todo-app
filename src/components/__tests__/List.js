import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import List from '../List';
import ListItem from '../ListItem';

describe('List Component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<List list={[]} visibility={''} />, div);
  });

  it('should filter all todo items', () => {
    const list = [
      { id: 'some-id-1', text: 'todo 1', isDone: false },
      { id: 'some-id-2', text: 'todo 2', isDone: true },
      { id: 'some-id-3', text: 'todo 3', isDone: false }
    ];
    const wrapper = mount(<List list={list} visibility={''} />);
    expect(wrapper.prop('list').length).toBe(3);

    wrapper.setProps({ visibility: 'whatever' });
    expect(wrapper.find(ListItem).length).toBe(3);
  });

  it('should filter completed todo items', () => {
    const list = [
      { id: 'some-id-1', text: 'todo 1', isDone: false },
      { id: 'some-id-2', text: 'todo 2', isDone: true },
      { id: 'some-id-3', text: 'todo 3', isDone: false }
    ];
    const wrapper = mount(<List list={list} visibility={'completed'} />);
    expect(wrapper.prop('list').length).toBe(3);
    expect(wrapper.find(ListItem).length).toBe(1);
  });

  it('should filter active todo items', () => {
    const list = [
      { id: 'some-id-1', text: 'todo 1', isDone: false },
      { id: 'some-id-2', text: 'todo 2', isDone: true },
      { id: 'some-id-3', text: 'todo 3', isDone: false }
    ];
    const wrapper = mount(<List list={list} visibility={'active'} />);
    expect(wrapper.prop('list').length).toBe(3);
    expect(wrapper.find(ListItem).length).toBe(2);
  });

  it('should be empty and hide clear completed button', () => {
    const tree = renderer.create(<List list={[]} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render two todo items', () => {
    const list = [
      { id: 'some-id-1', text: 'todo 1', isDone: false },
      { id: 'some-id-2', text: 'todo 2', isDone: false }
    ];
    const tree = renderer.create(<List list={list} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render one todo item and one completed', () => {
    const list = [
      { id: 'some-id-1', text: 'todo 1', isDone: false },
      { id: 'some-id-2', text: 'todo 2', isDone: true }
    ];
    const tree = renderer.create(<List list={list} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render two completed todo items and hide clear completed button', () => {
    const list = [
      { id: 'some-id-1', text: 'todo 1', isDone: true },
      { id: 'some-id-2', text: 'todo 2', isDone: true }
    ];
    const tree = renderer.create(<List list={list} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
