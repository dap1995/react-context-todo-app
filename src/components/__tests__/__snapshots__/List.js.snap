// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Component should be empty and hide clear completed button 1`] = `
<div
  className="sc-dnqmqq bVEpG"
>
  <div
    className="sc-iwsKbI NRpYV"
  />
  <div
    className="sc-gqjmRU dOQKBh"
  >
    <span
      className="sc-VigVT lbrqEn"
    >
      0 items left
    </span>
    <div
      className="sc-gZMcBi kgBKRm"
    >
      <a
        className="sc-htoDjs jCmNzT"
        href="#"
      >
        All
      </a>
      <a
        className="sc-htoDjs jCmNzT"
        href="#/active"
      >
        Active
      </a>
      <a
        className="sc-htoDjs jCmNzT"
        href="#/completed"
      >
        Completed
      </a>
    </div>
    <button
      className="sc-jTzLTM oYfQt"
      hidden={true}
      onClick={undefined}
    >
      Clear completed
    </button>
  </div>
</div>
`;

exports[`List Component should render one todo item and one completed 1`] = `
<div
  className="sc-dnqmqq bVEpG"
>
  <div
    className="sc-iwsKbI NRpYV"
  >
    <li
      className="sc-EHOje lhpJrP"
    >
      <button
        checked={false}
        className="sc-bZQynM bmWDER"
        onClick={[Function]}
      />
      <span
        checked={false}
        className="sc-ifAKCX cqwElN"
        onDoubleClick={[Function]}
      >
        todo 1
      </span>
      <button
        className="sc-gzVnrw dTBaCJ"
        onClick={[Function]}
      >
        X
      </button>
    </li>
    <li
      className="sc-EHOje lhpJrP"
    >
      <button
        checked={true}
        className="sc-bZQynM kWRFME"
        onClick={[Function]}
      />
      <span
        checked={true}
        className="sc-ifAKCX clBMZx"
        onDoubleClick={[Function]}
      >
        todo 2
      </span>
      <button
        className="sc-gzVnrw dTBaCJ"
        onClick={[Function]}
      >
        X
      </button>
    </li>
  </div>
  <div
    className="sc-gqjmRU dOQKBh"
  >
    <span
      className="sc-VigVT lbrqEn"
    >
      1 item left
    </span>
    <div
      className="sc-gZMcBi kgBKRm"
    >
      <a
        className="sc-htoDjs jCmNzT"
        href="#"
      >
        All
      </a>
      <a
        className="sc-htoDjs jCmNzT"
        href="#/active"
      >
        Active
      </a>
      <a
        className="sc-htoDjs jCmNzT"
        href="#/completed"
      >
        Completed
      </a>
    </div>
    <button
      className="sc-jTzLTM iqCtJN"
      hidden={false}
      onClick={undefined}
    >
      Clear completed
    </button>
  </div>
</div>
`;

exports[`List Component should render two completed todo items and hide clear completed button 1`] = `
<div
  className="sc-dnqmqq bVEpG"
>
  <div
    className="sc-iwsKbI NRpYV"
  >
    <li
      className="sc-EHOje lhpJrP"
    >
      <button
        checked={true}
        className="sc-bZQynM kWRFME"
        onClick={[Function]}
      />
      <span
        checked={true}
        className="sc-ifAKCX clBMZx"
        onDoubleClick={[Function]}
      >
        todo 1
      </span>
      <button
        className="sc-gzVnrw dTBaCJ"
        onClick={[Function]}
      >
        X
      </button>
    </li>
    <li
      className="sc-EHOje lhpJrP"
    >
      <button
        checked={true}
        className="sc-bZQynM kWRFME"
        onClick={[Function]}
      />
      <span
        checked={true}
        className="sc-ifAKCX clBMZx"
        onDoubleClick={[Function]}
      >
        todo 2
      </span>
      <button
        className="sc-gzVnrw dTBaCJ"
        onClick={[Function]}
      >
        X
      </button>
    </li>
  </div>
  <div
    className="sc-gqjmRU dOQKBh"
  >
    <span
      className="sc-VigVT lbrqEn"
    >
      0 items left
    </span>
    <div
      className="sc-gZMcBi kgBKRm"
    >
      <a
        className="sc-htoDjs jCmNzT"
        href="#"
      >
        All
      </a>
      <a
        className="sc-htoDjs jCmNzT"
        href="#/active"
      >
        Active
      </a>
      <a
        className="sc-htoDjs jCmNzT"
        href="#/completed"
      >
        Completed
      </a>
    </div>
    <button
      className="sc-jTzLTM iqCtJN"
      hidden={false}
      onClick={undefined}
    >
      Clear completed
    </button>
  </div>
</div>
`;

exports[`List Component should render two todo items 1`] = `
<div
  className="sc-dnqmqq bVEpG"
>
  <div
    className="sc-iwsKbI NRpYV"
  >
    <li
      className="sc-EHOje lhpJrP"
    >
      <button
        checked={false}
        className="sc-bZQynM bmWDER"
        onClick={[Function]}
      />
      <span
        checked={false}
        className="sc-ifAKCX cqwElN"
        onDoubleClick={[Function]}
      >
        todo 1
      </span>
      <button
        className="sc-gzVnrw dTBaCJ"
        onClick={[Function]}
      >
        X
      </button>
    </li>
    <li
      className="sc-EHOje lhpJrP"
    >
      <button
        checked={false}
        className="sc-bZQynM bmWDER"
        onClick={[Function]}
      />
      <span
        checked={false}
        className="sc-ifAKCX cqwElN"
        onDoubleClick={[Function]}
      >
        todo 2
      </span>
      <button
        className="sc-gzVnrw dTBaCJ"
        onClick={[Function]}
      >
        X
      </button>
    </li>
  </div>
  <div
    className="sc-gqjmRU dOQKBh"
  >
    <span
      className="sc-VigVT lbrqEn"
    >
      2 items left
    </span>
    <div
      className="sc-gZMcBi kgBKRm"
    >
      <a
        className="sc-htoDjs jCmNzT"
        href="#"
      >
        All
      </a>
      <a
        className="sc-htoDjs jCmNzT"
        href="#/active"
      >
        Active
      </a>
      <a
        className="sc-htoDjs jCmNzT"
        href="#/completed"
      >
        Completed
      </a>
    </div>
    <button
      className="sc-jTzLTM oYfQt"
      hidden={true}
      onClick={undefined}
    >
      Clear completed
    </button>
  </div>
</div>
`;
